packages/config-helpers/src/define-config.js

```diff
@@ -0,0 +1,380 @@
+/**
+ * @fileoverview defineConfig helper
+ * @author Nicholas C. Zakas
+ */
+
+//-----------------------------------------------------------------------------
+// Type Definitions
+//-----------------------------------------------------------------------------
+
+/** @typedef {import("eslint").Linter.Config} Config */
+/** @typedef {import("eslint").Linter.LegacyConfig} LegacyConfig */
+/** @typedef {import("eslint").ESLint.Plugin} Plugin */
+/** @typedef {import("eslint").Linter.RuleEntry} RuleEntry */
+/** @typedef {import("./types.ts").ExtendsElement} ExtendsElement */
+/** @typedef {import("./types.ts").SimpleExtendsElement} SimpleExtendsElement */
+/** @typedef {import("./types.ts").ConfigWithExtends} ConfigWithExtends */
+/** @typedef {import("./types.ts").ConfigWithExtendsArray} ConfigWithExtendsArray */
+
+//-----------------------------------------------------------------------------
+// Helpers
+//-----------------------------------------------------------------------------
+
+const eslintrcKeys = [
+	"env",
+	"extends",
+	"globals",
+	"ignorePatterns",
+	"noInlineConfig",
+	"overrides",
+	"parser",
+	"parserOptions",
+	"reportUnusedDisableDirectives",
+	"root",
+];
+
+/**
+ * Gets the name of a config object.
+ * @param {Config} config The config object.
+ * @param {string} indexPath The index path of the config object.
+ * @return {string} The name of the config object.
+ */
+function getConfigName(config, indexPath) {
+	if (config.name) {
+		return config.name;
+	}
+
+	return `UserConfig${indexPath}`;
+}
+
+/**
+ * Gets the name of an extension.
+ * @param {SimpleExtendsElement} extension The extension.
+ * @param {string} indexPath The index of the extension.
+ * @return {string} The name of the extension.
+ */
+function getExtensionName(extension, indexPath) {
+	if (typeof extension === "string") {
+		return extension;
+	}
+
+	if (extension.name) {
+		return extension.name;
+	}
+
+	return `ExtendedConfig${indexPath}`;
+}
+
+/**
+ * Determines if a config object is a legacy config.
+ * @param {Config|LegacyConfig} config The config object to check.
+ * @return {config is LegacyConfig} `true` if the config object is a legacy config.
+ */
+function isLegacyConfig(config) {
+	for (const key of eslintrcKeys) {
+		if (key in config) {
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Normalizes the plugin config by replacing the namespace with the plugin namespace.
+ * @param {string} userNamespace The namespace of the plugin.
+ * @param {Plugin} plugin The plugin config object.
+ * @param {Config} config The config object to normalize.
+ * @return {Config} The normalized config object.
+ */
+function normalizePluginConfig(userNamespace, plugin, config) {
+	// @ts-ignore -- ESLint types aren't updated yet
+	const pluginNamespace = plugin.meta?.namespace;
+
+	// don't do anything if the plugin doesn't have a namespace or rules
+	if (!pluginNamespace || (!config.rules && !config.processor)) {
+		return config;
+	}
```

Comment:
I think we can do the same early exit if `pluginNamespace === userNamespace` as well.